code/modules/research/mechanic/component_exchanger.dm

```diff
@@ -0,0 +1,256 @@
+//The newest tool in the Mechanic's arsenal, the Rapid Machinery Component Exchanger (RMCE)
+//It can load up a maximum of ten machinery modules and can replace machinery modules without even having to rebuild them
+//Useful if you want to replace a large amount of modules quickly and painlessly
+
+/obj/item/device/component_exchanger
+	name = "rapid machinery component exchanger"
+	desc = "A tool used to replace machinery components without having to deconstruct them. It can load up to ten components at once"
+	icon = 'icons/obj/device.dmi'
+	icon_state = "comp_exchanger"
+	gender = NEUTER
+	flags = FPRINT
+	slot_flags = SLOT_BELT
+	w_class = 2
+	item_state = "electronic"
+	m_amt = 0 //So the autolathe doesn't try to eat it
+	g_amt = 0
+	w_type = RECYK_ELECTRONIC
+	origin_tech = "magnets=2;engineering=4;materials=5;programming=3"
+	var/list/componentstorage = list()
+	var/maxcomponents = 10
+	var/emagged = 0 //So we can emag it for "improved" functionality
+
+//Do a quick check-up of what is loaded
+/obj/item/device/component_exchanger/attack_self(var/mob/user)
+
+	self_interact_menu(user)
+
+/obj/item/device/component_exchanger/attackby(var/obj/item/weapon/stock_parts/P, mob/user)
+
+	if(componentstorage.len < maxcomponents && istype(P)) //We have room
+		user.visible_message("<span class='notice'>[user] slots \a [P] into \the [src]'s component storage</span>", \
+		"<span class='notice'>You slot \a [P] into \the [src]'s component storage</span>")
+		user.drop_item(P)
+		componentstorage += P //Add it to the proper component storage list
+		playsound(get_turf(src), 'sound/items/Deconstruct.ogg', 50, 1) //User feedback
+	else
+		user << "<span class='warning'>\The [src] is full.</span>"
+
+//Redirect the attack only if it's a machine, otherwise don't bother
+/obj/item/device/component_exchanger/preattack(var/atom/A, var/mob/user, proximity_flag)
+
+	if(!proximity_flag)
+		return
+
+	if(istype(A, /obj/machinery))
+
+		var/obj/machinery/M = A
+
+		if(!M.panel_open)
+			user << "<span class='warning'>The maintenance hatch of \the [M] is closed, you can't just stab \the [src] into it and hope it'll work.</span>"
+			return
+
+		user.visible_message("<span class='notice'>[user] starts setting up \a [src] in \the [M]'s maintenance hatch</span>", \
+		"<span class='notice'>You carefully insert \the [src] through \the [M]'s maintenance hatch, it starts scanning the machine's components.</span>")
+
+		if(do_after(user, 30)) //3 seconds to obtain a complete reading of the machine's components
+
+			if(!proximity_flag)
```

Comment:
This won't work.
The prox flag is set when the click first happens. It checks at that instant, and it won't necessarily be the same after a do_after.
You need to use an actual Adjacent check here.